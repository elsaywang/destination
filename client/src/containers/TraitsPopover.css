:local(.traitPopover) {
    width: 20rem;
}

:local(.traitLink) {
    display: flex;
}

:local(.traitName) {
    display: block;
    text-overflow: ellipsis;
    overflow: hidden;
    white-space: nowrap;
    flex: 1;
}

:local(.traitSid) {
    margin-left: 0.5rem;
}
