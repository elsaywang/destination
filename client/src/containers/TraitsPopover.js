import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import OverlayTrigger from '@react/react-spectrum/OverlayTrigger';
import Popover from '@react/react-spectrum/Popover';
import Link from '@react/react-spectrum/Link';
import Button from '@react/react-spectrum/Button';
import ChevronDown from '@react/react-spectrum/Icon/ChevronDown';
import Wait from '@react/react-spectrum/Wait';
import { fetchTrait } from '../utils/fetchTrait';
import { maxVisibleTraits } from '../constants/includedInTraitsConstants';
import styles from './TraitsPopover.css';

class TraitsPopover extends Component {
    constructor() {
        super();

        this.state = {
            open: false,
            loading: true,
            traits: [],
            error: '',
        };
        this.onPopoverClick = this.onPopoverClick.bind(this);
    }

    static defaultProps = {
        maxVisibleTraits,
    };

    setStateAsync(state) {
        return new Promise(resolve => {
            this.setState(state, resolve);
        });
    }

    async onPopoverClick() {
        try {
            const responses = await Promise.all(
                this.props.sids.slice(0, this.props.maxVisibleTraits).map(sid => fetchTrait(sid)),
            );
            const successfulResponses = responses.filter(({ ok }) => ok === true);
            const traits = await Promise.all(successfulResponses.map(response => response.json()));

            await this.setStateAsync({
                loading: false,
                traits,
            });
        } catch (error) {
            await this.setStateAsync({
                loading: false,
                error,
            });
        }
    }

    renderAndMore = () => {
        const extraTraits = this.props.sids.length - this.state.traits.length;

        return extraTraits ? (
            <div className={styles.andMore}>
                And {this.props.sids.length - this.state.traits.length} more.
            </div>
        ) : null;
    };

    renderTraitLinks = () => {
        return this.state.traits.map(({ sid, name }) => {
            const label = `${name} - ${sid}`;
            const traitUrl = `/portal/Traits/Traits.ddx#view/${sid}`;

            return (
                <Link key={sid} href={traitUrl} className={styles.traitLink}>
                    <span className={styles.traitName}>{name}</span>
                    <span className={styles.traitSid}>({sid})</span>
                </Link>
            );
        });
    };

    renderContent = () => (
        <Fragment>
            {this.renderTraitLinks()}
            {this.renderAndMore()}
        </Fragment>
    );

    render() {
        const { sids } = this.props;
        const singular = 'Trait';
        const plural = 'Traits';
        const number = sids.length;
        const label = number + ' ' + (number === 1 ? singular : plural);

        return (
            <Fragment>
                <OverlayTrigger
                    trigger="click"
                    placement="bottom"
                    onShow={this.onPopoverClick}
                    data-test="overlay-trigger">
                    <Button
                        quiet
                        variant="action"
                        icon={<ChevronDown />}
                        data-test="overlay-trigger-button">
                        {label}
                    </Button>
                    <Popover
                        title={'Included in ' + label}
                        className={styles.traitPopover}
                        data-test="overlay-trigger-popover">
                        {this.state.loading || !this.state.traits.length ? (
                            <Wait />
                        ) : (
                            this.renderContent()
                        )}
                    </Popover>
                </OverlayTrigger>
            </Fragment>
        );
    }
}

TraitsPopover.propTypes = {
    sids: PropTypes.array.isRequired,
    maxVisibleTraits: PropTypes.number,
};

export default TraitsPopover;
