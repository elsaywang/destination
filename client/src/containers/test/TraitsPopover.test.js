import React from 'react';
import { shallow } from 'enzyme';
import TraitsPopover from '../TraitsPopover';
import OverlayTrigger from '@react/react-spectrum/OverlayTrigger';
import Popover from '@react/react-spectrum/Popover';
import Link from '@react/react-spectrum/Link';
import Button from '@react/react-spectrum/Button';
import Wait from '@react/react-spectrum/Wait';

describe('<TraitsPopover /> component', () => {
    describe('rendering', () => {
        it('matches snapshot', () => {
            const wrapper = shallow(<TraitsPopover sids={[1, 2, 3]} />);

            expect(wrapper).toMatchSnapshot();
        });

        it('renders <OverlayTrigger /> with <Button /> and <Popover /> children', () => {
            const wrapper = shallow(<TraitsPopover sids={[1, 2, 3]} />);

            const overlayTrigger = wrapper.find(OverlayTrigger);

            expect(overlayTrigger.filter('[data-test="overlay-trigger"]').exists()).toBe(true);
            expect(overlayTrigger.find(Button).exists()).toBe(true);
            expect(overlayTrigger.find(Popover).exists()).toBe(true);
        });

        it('renders <Button /> trigger with correct Trait(s) labeling', () => {
            const wrapper = shallow(<TraitsPopover sids={[1, 2]} />);

            expect(
                wrapper
                    .find(Button)
                    .filter('[data-test="overlay-trigger-button"]')
                    .html(),
            ).toContain('2 Traits');

            wrapper.setProps({
                sids: [1],
            });

            expect(
                wrapper
                    .find(Button)
                    .filter('[data-test="overlay-trigger-button"]')
                    .html(),
            ).toContain('1 Trait');
        });

        it('renders <Popover /> title with correct "Included in x Trait(s)" labeling', () => {
            const wrapper = shallow(<TraitsPopover sids={[1, 2]} />);

            expect(wrapper.find(Popover).props().title).toContain('Included in 2 Traits');

            wrapper.setProps({
                sids: [1],
            });

            expect(wrapper.find(Popover).props().title).toContain('Included in 1 Trait');
        });

        it('renders <Wait /> when loading', () => {
            const wrapper = shallow(<TraitsPopover sids={[1, 2, 3]} />);

            wrapper.setState({
                loading: true,
            });

            expect(wrapper.find(Wait).exists()).toBe(true);
        });

        it('renders content in Popover when not loading', () => {
            const wrapper = shallow(<TraitsPopover sids={[1, 2, 3]} />);

            wrapper.setState({
                loading: false,
                traits: [{ sid: 1, name: 'i am a trait' }],
            });

            const content = wrapper
                .find(Popover)
                .filter('[data-test="overlay-trigger-popover"]')
                .html();

            expect(wrapper.find(Wait).exists()).toBe(false);
            expect(
                wrapper
                    .find(Popover)
                    .filter('[data-test="overlay-trigger-popover"]')
                    .find(Link).length,
            ).toBe(1);
            expect(content).toContain('i am a trait');
            expect(content).toContain('(1)');
        });

        it('renders `maxVisibleTraits` traits and an "and more" message when many sids are provided', () => {
            const wrapper = shallow(<TraitsPopover sids={[1, 2]} maxVisibleTraits={1} />);

            wrapper.setState({
                loading: false,
                traits: [{ sid: 1, name: 'Trait 1' }],
            });

            expect(
                wrapper
                    .find(Popover)
                    .filter('[data-test="overlay-trigger-popover"]')
                    .find(Link).length,
            ).toBe(1);

            const content = wrapper
                .find(Popover)
                .filter('[data-test="overlay-trigger-popover"]')
                .html();

            expect(content).toContain('And 1 more.');
        });
    });
});
