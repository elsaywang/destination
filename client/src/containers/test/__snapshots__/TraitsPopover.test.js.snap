// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TraitsPopover /> component rendering matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TraitsPopover
    maxVisibleTraits={8}
    sids={
        Array [
            1,
            2,
            3,
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <OverlayTrigger
        boundariesElement={[Function]}
        crossOffset={0}
        data-test="overlay-trigger"
        defaultShow={false}
        disabled={false}
        flip={true}
        offset={0}
        onShow={[Function]}
        placement="bottom"
        trigger="click"
>
        <Button
                autoFocus={false}
                block={false}
                data-test="overlay-trigger-button"
                disabled={false}
                element="button"
                icon={<ChevronDown />}
                invalid={false}
                label=""
                logic={false}
                quiet={true}
                selected={false}
                variant="action"
        >
                3 Traits
        </Button>
        <Popover
                className={undefined}
                data-test="overlay-trigger-popover"
                open={true}
                placement="bottom"
                title="Included in 3 Traits"
                trapFocus={true}
                variant="default"
        >
                <Wait />
        </Popover>
</OverlayTrigger>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "boundariesElement": [Function],
        "children": Array [
          <Button
            autoFocus={false}
            block={false}
            data-test="overlay-trigger-button"
            disabled={false}
            element="button"
            icon={<ChevronDown />}
            invalid={false}
            label=""
            logic={false}
            quiet={true}
            selected={false}
            variant="action"
>
            3 Traits
</Button>,
          <Popover
            className={undefined}
            data-test="overlay-trigger-popover"
            open={true}
            placement="bottom"
            title="Included in 3 Traits"
            trapFocus={true}
            variant="default"
>
            <Wait />
</Popover>,
        ],
        "crossOffset": 0,
        "data-test": "overlay-trigger",
        "defaultShow": false,
        "disabled": false,
        "flip": true,
        "offset": 0,
        "onShow": [Function],
        "placement": "bottom",
        "trigger": "click",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "autoFocus": false,
            "block": false,
            "children": "3 Traits",
            "data-test": "overlay-trigger-button",
            "disabled": false,
            "element": "button",
            "icon": <ChevronDown />,
            "invalid": false,
            "label": "",
            "logic": false,
            "quiet": true,
            "selected": false,
            "variant": "action",
          },
          "ref": null,
          "rendered": "3 Traits",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Wait />,
            "className": undefined,
            "data-test": "overlay-trigger-popover",
            "open": true,
            "placement": "bottom",
            "title": "Included in 3 Traits",
            "trapFocus": true,
            "variant": "default",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <OverlayTrigger
          boundariesElement={[Function]}
          crossOffset={0}
          data-test="overlay-trigger"
          defaultShow={false}
          disabled={false}
          flip={true}
          offset={0}
          onShow={[Function]}
          placement="bottom"
          trigger="click"
>
          <Button
                    autoFocus={false}
                    block={false}
                    data-test="overlay-trigger-button"
                    disabled={false}
                    element="button"
                    icon={<ChevronDown />}
                    invalid={false}
                    label=""
                    logic={false}
                    quiet={true}
                    selected={false}
                    variant="action"
          >
                    3 Traits
          </Button>
          <Popover
                    className={undefined}
                    data-test="overlay-trigger-popover"
                    open={true}
                    placement="bottom"
                    title="Included in 3 Traits"
                    trapFocus={true}
                    variant="default"
          >
                    <Wait />
          </Popover>
</OverlayTrigger>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "boundariesElement": [Function],
          "children": Array [
            <Button
              autoFocus={false}
              block={false}
              data-test="overlay-trigger-button"
              disabled={false}
              element="button"
              icon={<ChevronDown />}
              invalid={false}
              label=""
              logic={false}
              quiet={true}
              selected={false}
              variant="action"
>
              3 Traits
</Button>,
            <Popover
              className={undefined}
              data-test="overlay-trigger-popover"
              open={true}
              placement="bottom"
              title="Included in 3 Traits"
              trapFocus={true}
              variant="default"
>
              <Wait />
</Popover>,
          ],
          "crossOffset": 0,
          "data-test": "overlay-trigger",
          "defaultShow": false,
          "disabled": false,
          "flip": true,
          "offset": 0,
          "onShow": [Function],
          "placement": "bottom",
          "trigger": "click",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "autoFocus": false,
              "block": false,
              "children": "3 Traits",
              "data-test": "overlay-trigger-button",
              "disabled": false,
              "element": "button",
              "icon": <ChevronDown />,
              "invalid": false,
              "label": "",
              "logic": false,
              "quiet": true,
              "selected": false,
              "variant": "action",
            },
            "ref": null,
            "rendered": "3 Traits",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <Wait />,
              "className": undefined,
              "data-test": "overlay-trigger-popover",
              "open": true,
              "placement": "bottom",
              "title": "Included in 3 Traits",
              "trapFocus": true,
              "variant": "default",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
